## Flask Projects with Web and RESTful APIs 

### 1. **To-Do App**

#### RESTful API Features:
- **GET /todos**: Retrieve all to-do items.
- **POST /todos**: Add a new to-do item.
- **PUT /todos/<id>**: Update a specific to-do item.
- **DELETE /todos/<id>**: Delete a specific to-do item.
- **GET /todos/<id>**: Retrieve a specific to-do item.

#### Model:
- **Todo**: Represents a single to-do item.
  - `id`: Unique identifier for the to-do.
  - `task`: Description of the to-do.
  - `is_completed`: Boolean indicating whether the to-do is completed.


---
### Reference Videos

1. Python Flask REST API tutorial part 1 [Revisory Notes on Python Flask Routes](../notes/revisory_notes/python_flask_routes.ipynb)  
      * video: https://us06web.zoom.us/rec/share/-CpiN2P8ItHZmkyTJ165CbF-jlf7h56DhJpeWWjZjfAdwAZ9TVDHtnC2oXz7k0zj.VxD_UCVEyceD8yTm?startTime=1735761858000
      * Passcode: Passcode: Uj+G8%ik  

2.  Python Flask REST API tutorial part 2 [Revisory Notes on Python Flask Routes](../notes/revisory_notes/python_flask_routes.ipynb)  
      * video: https://us06web.zoom.us/rec/share/-CpiN2P8ItHZmkyTJ165CbF-jlf7h56DhJpeWWjZjfAdwAZ9TVDHtnC2oXz7k0zj.VxD_UCVEyceD8yTm?startTime=1735762871000  
      * Passcode:  Uj+G8%ik  

3.  Python Flask REST API tutorial part 3 [Revisory Notes on Python Flask Routes](../notes/revisory_notes/python_flask_routes.ipynb)  
      * video: https://us06web.zoom.us/rec/share/y0xKk6FaiDoyey1r6QEJ8QZPgz91SYjt_9fA0fXLkQxWhWbxgJTz4YzVAY-LLcys.wEyxQkm5JU8agG7C?startTime=1735842995000  
      * Passcode:  b=^pD+1A

4.  Python Flask REST API tutorial part 4 [Revisory Notes on Python Flask Routes](../notes/revisory_notes/python_flask_routes.ipynb)  
      * video: https://us06web.zoom.us/rec/share/y0xKk6FaiDoyey1r6QEJ8QZPgz91SYjt_9fA0fXLkQxWhWbxgJTz4YzVAY-LLcys.wEyxQkm5JU8agG7C?startTime=1735847140000  
      * Passcode:  b=^pD+1A

5.  Python Flask REST API tutorial part 5 [Revisory Notes on Python Flask Routes](../notes/revisory_notes/python_flask_routes.ipynb)  
      * video: https://us06web.zoom.us/rec/share/y0xKk6FaiDoyey1r6QEJ8QZPgz91SYjt_9fA0fXLkQxWhWbxgJTz4YzVAY-LLcys.wEyxQkm5JU8agG7C?startTime=1735848942000 
      * Passcode:  b=^pD+1A

In [ ]:
!pip install flask --break-system-packages
!pip install flask-restx  --break-system-packages

In [ ]:
from http import HTTPStatus
import nest_asyncio
from flask import Flask, Blueprint, Response, jsonify, request
from werkzeug.serving import run_simple

# Allow Flask to run in a Jupyter Notebook
nest_asyncio.apply()

# Initialize Flask app
app = Flask(__name__)

# In-memory storage for to-do tasks
todos = {} 
""" 
 todos = {
     "1": {
         'id': 1,
         'task': " perform daily task number 1",
         'is_completed': True
     },
     
     "2": {
         'id': 2,
         'task': " perform daily task number 2",
         'is_completed': True
     }
 }
 todos.keys() = partiallist<"1", "2">
 list( todos.keys()) = ["1", "2"]
 todos.values() = partialist<{
         'id': 1,
         'task': " perform daily task number 1",
         'is_completed': True
     }, 
     {
         'id': 2,
         'task': " perform daily task number 2",
         'is_completed': True
     }
     >
 list( todos.values()) = [{
         'id': 1,
         'task': " perform daily task number 1",
         'is_completed': True
     }, 
     {
         'id': 2,
         'task': " perform daily task number 2",
         'is_completed': True
     }]
"""

# Define a Blueprint for the to-do routes
todo_bp = Blueprint("todo", __name__)

# - **GET /todos**: Retrieve all to-do items.
# - **POST /todos**: Add a new to-do item.
# - **PUT /todos/<id>**: Update a specific to-do item.
# - **DELETE /todos/<id>**: Delete a specific to-do item.
# - **GET /todos/<id>**: Retrieve a specific to-do item.
# CRUD = CREATE[POST] READ[GET] UPDATE[PUT] DELETE[DELETE]

# READ =   GET /todos, 
@todo_bp.route('/todos', methods=['GET'])
def get_todos() -> tuple[Response, int]:
    items = list(todos.values())
    response_data = jsonify(items)
    response_status_code = HTTPStatus.OK
    
    return response_data, response_status_code

    # return jsonify(list(todos.values())), HTTPStatus.OK

# CREATE = POST /todos
@todo_bp.route('/todos', methods=['POST'])
def create_todo()-> tuple[Response, int]:
    data = request.get_json()
    if not data or 'task' not in data:
        response_data = jsonify({'error': 'Task is required'})
        response_status_code = HTTPStatus.BAD_REQUEST
        
        return response_data, response_status_code
    # return jsonify({'error': 'Task is required'}), HTTPStatus.BAD_REQUEST
    
    new_id = max(todos.keys(), default=0) + 1 
    new_todo = {
       'id': new_id,
       'task': data['task'],
       'is_completed': False
    }
    todos[new_id] = new_todo
    
    response_data = jsonify(new_todo)
    response_status_code = HTTPStatus.CREATED
    
    return response_data, response_status_code
    

# READ =  /todos/<id>
@todo_bp.route('/todos/<int:id>', methods=['GET'])
def get_todo(id: int)-> tuple[Response, int]:
    # if id not in todos: # prone to generating error...
    #     response_data = jsonify({'error', 'TO-DO not found'})
    #     response_status_code = HTTPStatus.NOT_FOUND
        
    #     return response_data, response_status_code
        # return  jsonify({'error', 'TO-DO not found'}), HTTPStatus.NOT_FOUND
    # x = 1  
    # not x = None
    # if not x: check if x is None
    todo = todos.get(id) # value or None e.g todos.get(1)... , todos.get("3") None  todos[id]
    # if todo is None: expensive will use two memory spaces # since it is checking equality of two types
    # if todo == None: expensive will use two memory spaces # since it is checking equality of two values
    if not todo: # if the todo is None    # will only use one memory space since it is checking for type
        response_data = jsonify({'error', 'TO-DO not found'})
        response_status_code = HTTPStatus.NOT_FOUND
        
        return response_data, response_status_code
        # return  jsonify({'error', 'TO-DO not found'}), HTTPStatus.NOT_FOUND
    
    response_data = jsonify(todo)
    response_status_code = HTTPStatus.OK 
    
    return response_data, response_status_code
    
    

# UPDATE = PUT /todos/<id>
@todo_bp.route('/todos/<int:id>', methods=['PUT'])
def update_todo(id: int)-> tuple[Response, int]:
    todo = todos.get(id) 
    """
     id, task, is_complete 
    """
    if not todo: 
        response_data = jsonify({'error', 'TO-DO not found'})
        response_status_code = HTTPStatus.NOT_FOUND
        
        return response_data, response_status_code
    
    data = request.get_json()
    """ 
    task | is_complete
    """
    if not data:
        response_data = jsonify({'error': 'Invalid data'})
        response_status_code = HTTPStatus.BAD_REQUEST
    
    old_task =  todo.get('task')
    old_is_complete = todo.get('is_complete')
    
    new_task = data.get('task', old_task) # if None use old_task
    new_is_complete = data.get('is_complete', old_is_complete) # if None use old_is_complete
    
    todo.update({
        'task': new_task,
        'is_complete': new_is_complete
    })
    
    todos[id] = todo
    
    response_data = jsonify(todo)
    response_status_code = HTTPStatus.OK
    
    return response_data, response_status_code
    
    

# DELETE = /todos/<id>
@todo_bp.route('/todos/<int:id>', methods=['DELETE'])
def delete_todo(id: int)-> tuple[Response, int]:
    todo = todos.get(id) 
    if not todo: 
        response_data = jsonify({'error', 'TO-DO not found'})
        response_status_code = HTTPStatus.NOT_FOUND
        
        return response_data, response_status_code
    
    deleted_todo =  todos.pop(id)
    
    response_data = jsonify(deleted_todo)  
    response_status_code = HTTPStatus.NO_CONTENT
    
    return response_data, response_status_code

  
# Register the Blueprint  with the Flask app  
app.register_blueprint(todo_bp)

# localhost = 127.0.0.1, 0.0.0.0
run_simple('localhost', 5000, app, use_debugger=True)

 * Running on http://localhost:5000
Press CTRL+C to quit


127.0.0.1 - - [17/Jan/2025 22:05:26] "POST /todos HTTP/1.1" 201 -
127.0.0.1 - - [17/Jan/2025 22:05:26] "POST /todos HTTP/1.1" 201 -
127.0.0.1 - - [17/Jan/2025 22:05:34] "DELETE /todos/2 HTTP/1.1" 204 -
127.0.0.1 - - [17/Jan/2025 22:06:08] "DELETE /todos/1 HTTP/1.1" 204 -
127.0.0.1 - - [17/Jan/2025 22:06:46] "POST /todos HTTP/1.1" 201 -
127.0.0.1 - - [17/Jan/2025 22:06:46] "POST /todos HTTP/1.1" 201 -
127.0.0.1 - - [17/Jan/2025 22:06:49] "GET /todos HTTP/1.1" 200 -
127.0.0.1 - - [17/Jan/2025 22:06:53] "DELETE /todos/1 HTTP/1.1" 204 -
127.0.0.1 - - [17/Jan/2025 22:06:53] "GET /todos HTTP/1.1" 200 -


# Flask Projects with Web and RESTful APIs

## 1. To-Do App

### RESTful API Features

#### 1. Retrieve All To-Do Items
- **GET `/todos`**
  - **Description**: Retrieves all to-do items.
  - **Response Schema**:
    ```json
    {
      "success": true,
      "data": [
        {
          "id": 1,
          "task": "Buy groceries",
          "is_completed": false
        },
        {
          "id": 2,
          "task": "Call the plumber",
          "is_completed": true
        }
      ]
    }
    ```
  - **HTTP Status Codes**:
    - `200 OK`: Successfully retrieved all to-do items.
    - `500 Internal Server Error`: Server issue occurred.

---

#### 2. Add a New To-Do Item
- **POST `/todos`**
  - **Description**: Creates a new to-do item.
  - **Request Schema**:
    ```json
    {
      "task": "Clean the house",
      "is_completed": false
    }
    ```
  - **Response Schema**:
    ```json
    {
      "success": true,
      "data": {
        "id": 3,
        "task": "Clean the house",
        "is_completed": false
      }
    }
    ```
  - **HTTP Status Codes**:
    - `201 Created`: To-do item successfully created.
    - `400 Bad Request`: Invalid input.
    - `500 Internal Server Error`: Server issue occurred.

---

#### 3. Update a Specific To-Do Item
- **PUT `/todos/<id>`**
  - **Description**: Updates the details of a specific to-do item by its `id`.
  - **Path Parameter**:
    - `id` (integer): Unique identifier for the to-do item.
  - **Request Schema**:
    ```json
    {
      "task": "Buy groceries and cook dinner",
      "is_completed": true
    }
    ```
  - **Response Schema**:
    ```json
    {
      "success": true,
      "data": {
        "id": 1,
        "task": "Buy groceries and cook dinner",
        "is_completed": true
      }
    }
    ```
  - **HTTP Status Codes**:
    - `200 OK`: To-do item successfully updated.
    - `400 Bad Request`: Invalid input.
    - `404 Not Found`: To-do item not found.
    - `500 Internal Server Error`: Server issue occurred.

---

#### 4. Delete a Specific To-Do Item
- **DELETE `/todos/<id>`**
  - **Description**: Deletes a to-do item by its `id`.
  - **Path Parameter**:
    - `id` (integer): Unique identifier for the to-do item.
  - **Response Schema**:
    ```json
    {
      "success": true,
      "message": "To-do item deleted successfully."
    }
    ```
  - **HTTP Status Codes**:
    - `204 No Content`: To-do item successfully deleted.
    - `404 Not Found`: To-do item not found.
    - `500 Internal Server Error`: Server issue occurred.

---

#### 5. Retrieve a Specific To-Do Item
- **GET `/todos/<id>`**
  - **Description**: Retrieves a specific to-do item by its `id`.
  - **Path Parameter**:
    - `id` (integer): Unique identifier for the to-do item.
  - **Response Schema**:
    ```json
    {
      "success": true,
      "data": {
        "id": 2,
        "task": "Call the plumber",
        "is_completed": true
      }
    }
    ```
  - **HTTP Status Codes**:
    - `200 OK`: To-do item successfully retrieved.
    - `404 Not Found`: To-do item not found.
    - `500 Internal Server Error`: Server issue occurred.

---

### Model
- **Todo**: Represents a single to-do item.
  - `id`: Unique identifier for the to-do (integer).
  - `task`: Description of the to-do (string).
  - `is_completed`: Boolean indicating whether the to-do is completed (boolean).  


## *N.B:* Use mysql database for the model
### Tutorials
1.  Python Flask REST API WITH FLASK RESTX tutorial part 1 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/dpBbtagnvaXHTRnMJRcAEFQ0YndQnv_HmccmpuAySEw4Svc98Q7lt-9Oky2JyRXF.4I3g4VIylctYvP1I?startTime=1735932088000
      * Passcode: G$JRMk6T

2.  Python Flask REST API WITH FLASK RESTX tutorial part 2 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/yuQmqXR17qWRDz1ORqruqBEw-9P5mKvLNI1u30TI12nZJ5z9GuG11f2P3w0hToTI.BKACjrs4MPB0AIjX?startTime=1736361656000
      * Passcode: D&z%m=T3

3.  Python Flask REST API WITH FLASK RESTX tutorial part 3 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/yuQmqXR17qWRDz1ORqruqBEw-9P5mKvLNI1u30TI12nZJ5z9GuG11f2P3w0hToTI.BKACjrs4MPB0AIjX?startTime=1736362994000
      * Passcode:  D&z%m=T3

4.  Python Flask REST API WITH FLASK RESTX tutorial part 4 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/yuQmqXR17qWRDz1ORqruqBEw-9P5mKvLNI1u30TI12nZJ5z9GuG11f2P3w0hToTI.BKACjrs4MPB0AIjX?startTime=1736365139000
      * Passcode: D&z%m=T3

5.  Python Flask REST API WITH FLASK RESTX tutorial part 5 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/i5ezzwX5aOg4JvgsMLV0fC_s5J-AkINvEiXRe593fB22fmt_VXADylP851A17fbN.zQJPilVsq5wd3d2E?startTime=1736451425000
      * Passcode: ha^Rr3C^

6. Python Flask REST API WITH FLASK RESTX tutorial part 6 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/u0iTj1B5L8yeJcWG3DQ7QfeFGEvEef6Afs9IFctJQ5WUCc16utMmhQKQmHfmq7Y0.yLcdo-VtqOFYAyIc?startTime=1736536043000
      * Passcode: p@rx+hs6  

7. Python Flask REST API WITH FLASK RESTX tutorial part 7 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/u0iTj1B5L8yeJcWG3DQ7QfeFGEvEef6Afs9IFctJQ5WUCc16utMmhQKQmHfmq7Y0.yLcdo-VtqOFYAyIc?startTime=1736539798000  
      * Passcode: p@rx+hs6  

8. Python Flask REST API WITH FLASK RESTX tutorial part 8 [Revisory Notes on Python Flask Routes](./notes/revisory_notes/python_flaskx_routes.ipynb)
      * video: https://us06web.zoom.us/rec/share/u0iTj1B5L8yeJcWG3DQ7QfeFGEvEef6Afs9IFctJQ5WUCc16utMmhQKQmHfmq7Y0.yLcdo-VtqOFYAyIc?startTime=1736540922000  
      * Passcode:  p@rx+hs6  

In [ ]:
%pip install flask --break-system-packages
%pip install flask-restx  --break-system-packages
%pip install flask flask-sqlalchemy   --break-system-packages
%pip install  mysql-connector-python --break-system-packages

In [ ]:
import nest_asyncio
from werkzeug.serving import run_simple
from flask import Flask
from flask_restx import Api, Namespace, Resource, fields
from http import HTTPStatus

nest_asyncio.apply()

# Initialize Flask app and API
app = Flask(__name__)

# Wrap the initialized app with Flask-Restx swagger
api  = Api(app, version="1.0", title="", description="")

# In-memory storage
 
# In-memory storage for to-do tasks
todos = {} 



    

# localhost = 127.0.0.1, 0.0.0.0
run_simple('localhost', 5000, app, use_debugger=True)  

In [ ]:
%pip install flask --break-system-packages
%pip install flask-restx  --break-system-packages
%pip install flask flask-sqlalchemy   --break-system-packages
%pip install  mysql-connector-python --break-system-packages


import nest_asyncio
from werkzeug.serving import run_simple
from flask import Flask
from flask_restx import Api, Namespace, Resource, fields
from http import HTTPStatus

nest_asyncio.apply()

# Initialize Flask app and API
app = Flask(__name__)

# Wrap the initialized app with Flask-Restx swagger
api  = Api(app, version="1.0", title="", description="")
 
# Use RDBMS=> MYSQL
    

# localhost = 127.0.0.1, 0.0.0.0
run_simple('localhost', 5000, app, use_debugger=True)  